### ISSUE TITLE
build_pptx() 

### ISSUE THREAD

**Issue by brshallo** [jhelvy/renderthis#4]
Picking-up from [twitter conversation](https://twitter.com/JohnHelveston/status/1355581730053185539?s=20) and having jhelvy/xaringanBuilder be place to collect functionality on outputting xaringan output formats (per conversation with @EvaMaeRey ). 

A functionality that is in EvaMaeRey/xaringanPrinter and not here is in outputting to pptx format. I pasted the function below (in it's [other location](https://github.com/EvaMaeRey/xaringanPrinter/blob/master/R/conversion.R) it had the title `slides_pptx()`).

A few notes:
* Function is in draft form (there are currently no checks on file paths etc.)
* May want to handle printing differently (i.e. reliance on `pagedown::chrome_print()`), particularly depending on outcome from #1
* Adding function would probably mean needing to add {officer} to suggests in the DESCRIPTION
* EvaMaeRey/flipbookr#33 is where initial notes are regarding considerations on this function

Feel free to make any changes. Let me know your thoughts and if you'd like me to open a PR?

```r
#' Print Flipbook to Powerpoint
#'
#' @param input Name of the Rmd or html slides file.
#' @return A .pdf, .html, series of .png images, and a .pptx
#' @export
build_pptx <- function(input){

  filepath_no_ext <- fs::path_ext_remove(input)
  filepath_pdf <- fs::path_ext_set(filepath_no_ext, "pdf")
  filepath_pptx <- fs::path_ext_set(filepath_no_ext, "pptx")

  pagedown::chrome_print(input) # run if have not already created pdf

  screen_imgs <- pdftools::pdf_convert(pdf = filepath_pdf, format = "png", verbose = FALSE)

  doc <- officer::read_pptx()
  for (i in 1:length(screen_imgs)) {
    doc <- doc %>%
      officer::add_slide(layout = "Blank", master = "Office Theme") %>%
      officer::ph_with(
        value = officer::external_img(screen_imgs[[i]]),
        location = officer::ph_location_fullsize()
      )
  }

  print(doc, filepath_pptx)
}
```

*Fixed a typo after initial post -- had `input` where meant to have `screen_imgs` when initiating `for` loop.*

**Comment by jhelvy** [comment]
This is awesome! If I'm understanding this function correctly, it looks like it creates a png file for each slide, then inserts those pngs into a pptx file with {officer}. 

If that is correct, I might suggest matching the structure of `build_gif()`. It has lines for checking input types, and it uses `build_pdf()` to build the pdf if the input is a .Rmd or .html (and the pdf is missing). It then uses `magick::image_convert()` to convert the pdf into a list of png images. So you could create a similar structure here to get the list of pngs and then loop through the list to make the pptx. And I agree on putting {officer} in the suggests rather than imports, I'd just add an `if` statement at the top of the function to make sure it's installed (e.g. see the top of the `xaringan_to_pdf()` function). 

Let me know if I'm understanding this correctly. Happy to review a PR on this!

**Comment by brshallo** [comment]
Thanks for the explicit tips, were very helpful! I think I've almost got it and will open a PR soon. The part I'm stuck on is in looping through the images.

I can't get the approach using `magick` to work with the pngs (ran into this when first writing function as well...):

Specifically if I try and replace the `pdftools::pdf_convert()` step with the following:
```r
    pdf <- magick::image_read(input, density = density)
    pngs <- magick::image_convert(pdf, 'png')
```

I get an error with `officer::ph_with()` when looping through:

```
Error in UseMethod("ph_with", value) : 
  no applicable method for 'ph_with' applied to an object of class "magick-image"
```

The approach here (using `pdftools`) produces the correct powerpoint output:

```r

build_pptx <- function(input, output_file = NULL) {
  
  if (!requireNamespace("officer", quietly = TRUE)) {
    stop("`officer` is required: install.packages('officer')")
  }
  
  # Remove if can get working with magick
  if (!requireNamespace("pdftools", quietly = TRUE)) {
    stop("`officer` is required: install.packages('pdftools')")
  }
  
  assert_path_ext(input, c("rmd", "html", "pdf"))
  input <- fs::path_abs(input)
  
  if (test_path_ext(input, c("rmd", "html"))) {
    build_pdf(input, output_file)
    input <- fs::path_ext_set(input, "pdf")
  }
  
  if (is.null(output_file)) {
    output_file <- fs::path_ext_set(input, "pptx")
  } else if (test_path_ext(output_file, "pptx")) {
    stop("`output_file` should be NULL or have .pptx extension")
  }
  cli::cli_process_start(
    "Building {.file {fs::path_file(output_file)}} from {.path {fs::path_file(input)}}",
    on_exit = "done"
  )
  
  pngs <- pdftools::pdf_convert(pdf = input, 
                                format = "png", 
                                verbose = FALSE)
  
  doc <- officer::read_pptx()
  for (i in 1:length(pngs)) {
    doc <- doc %>%
      officer::add_slide(layout = "Blank", master = "Office Theme") %>%
      officer::ph_with(
        value = officer::external_img(pngs[[i]]),
        location = officer::ph_location_fullsize()
      )
  }
  
  print(doc, output_file)
}

```

However:

* It would nice if we could use `magick` here to be consistent (and so could also include `density` argument -- `pdftools::pdf_convert()` allows a `dpi` argument)
* Also the above approach saves .png's of each page in your folder (cluttering it up). It would be fine if these went to either a temp or say a folder named after `input` but not sure how best to specify that ( ropensci/pdftools#73 )

@jhelvy any ideas on this?

**Comment by jhelvy** [comment]
The `officer::external_img()` function is looking for a path to an image saved somewhere, so my (hacky) solution here was to use`magick::image_write()` to save each png to a temporary file and then use the path to that temp file inside `officer::external_img()`. Looks like it's working for me. 

```r
build_pptx <- function(input, output_file = NULL, density = "72x72") {
    if (!requireNamespace("officer", quietly = TRUE)) {
        stop("`officer` is required: install.packages('officer')")
    }

    assert_path_ext(input, c("rmd", "html", "pdf"))
    input <- fs::path_abs(input)

    if (test_path_ext(input, c("rmd", "html"))) {
        build_pdf(input, output_file)
        input <- fs::path_ext_set(input, "pdf")
    }

    if (is.null(output_file)) {
        output_file <- fs::path_ext_set(input, "pptx")
    } else if (test_path_ext(output_file, "pptx")) {
        stop("`output_file` should be NULL or have .pptx extension")
    }
    cli::cli_process_start(
        "Building {.file {fs::path_file(output_file)}} from {.path {fs::path_file(input)}}",
        on_exit = "done"
    )
    pdf <- magick::image_read(input, density = density)
    pngs <- magick::image_convert(pdf, 'png')

    doc <- officer::read_pptx()
    for (i in 1:length(pngs)) {
        png_path <- magick::image_write(
            pngs[i], tempfile(fileext = ".png"))
        doc <- officer::add_slide(
            doc,
            layout = "Blank",
            master = "Office Theme"
        )
        doc <- officer::ph_with(
            doc,
            value = officer::external_img(png_path),
            location = officer::ph_location_fullsize()
        )
    }

  print(doc, output_file)
}
```

**Comment by jhelvy** [comment]
Solved with #5
